content/umdctf_2024/ready_aim_fire.md

```diff
@@ -0,0 +1,161 @@
+            }
+        }
+        if (bufIndex >= 32) {
+            throw out_of_range{""};
+        }
+    }
+};
+
+void fire_weapon() {
+    Cannon w;
+    w.fire();
+}
+
+int main() {
+    int target_assist;
+    cout << "Quick! While the spice harvester's shields are down! Fire the laser cannon!" << endl;
+    cout << &target_assist << endl;
+
+    try {
+        fire_weapon();
+        cout << "Looks like you missed your opportunity to fire." << endl;
+    }
+    catch (exception e) {
+        cout << "Seems like you missed." << endl;
+    }
+}
+```
+
+We can see that there is a constraint though, if we do a BOF, it will trigger an exception, which will be caught in the main function. Additionally, I noticed an extra function called 'direct_hit,' which suggested that we should attempt to redirect the exception thrown during a BOF into the catch block of 'direct_hit()' to print the flag (the intended solution). Unfortunately, my implementation didn't work as expected. Therefore, my next idea was to overwrite the return address of the main function with the address of 'print_flag()' after the exception was triggered
+
+### Solution
+
+#### Unintented
+
+When first trying to do the BOF the program crashed in the exception handler and we could clearly see that we had overwritten there something.
+
+```c
+0x00007ffd8a60ad20│+0x0000: "XXXXXXXX+'@"        ← $rsp, $rbp   <---------------- crash 
+0x00007ffd8a60ad28│+0x0008: 0x000000000040272b  →  <main+270> mov rax, rbx  
+0x00007ffd8a60ad30│+0x0010: 0x004023e630303030
+0x00007ffd8a60ad38│+0x0018: 0x004023e600000000
+0x00007ffd8a60ad40│+0x0020: 0x004023e600000000
+0x00007ffd8a60ad48│+0x0028: 0x004023e600000000
+0x00007ffd8a60ad50│+0x0030: 0x004023e600000000
+0x00007ffd8a60ad58│+0x0038: 0x004023e600000000
+─────────────────────────────────────────────────────────────────────────────────────────────────────────────── code:x86:64 ────
+     0x40283a <std::exception::exception(std::exception const&)+12> mov    QWORD PTR [rbp-0x10], rsi
+     0x40283e <std::exception::exception(std::exception const&)+16> lea    rdx, [rip+0x255b]        # 0x404da0 <_ZTVSt9exception@GLIBCXX_3.4+16>
+     0x402845 <std::exception::exception(std::exception const&)+23> mov    rax, QWORD PTR [rbp-0x8]
+●→   0x402849 <std::exception::exception(std::exception const&)+27> mov    QWORD PTR [rax], rdx
+     0x40284c <std::exception::exception(std::exception const&)+30> nop    
+     0x40284d <std::exception::exception(std::exception const&)+31> pop    rbp
+     0x40284e <std::exception::exception(std::exception const&)+32> ret    
+     0x40284f                  nop    
+     0x402850 <Cannon::Cannon()+0> endbr64 
+
+```
+
+To address this issue, I extended my payload beyond the stack leak address by applying an offset to ensure it wouldn't interfere with the existing code. Then I continued with the address found on the stack ensuring that the exception would execute smoothly and return to the main function.
```

Comment:
here I don't fully understand what _extending beyond the stack leak_ and what _applying an offset_ mean. might be a skill issue tho lol